File: RoboLabWP/RoboLabWP/Model/Game/MZGame.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using RoboLabWP.Model.Basic_Types;

//
//  MZGame.cs
//  Maze
//
//  Класс данных текущей игры, у которого никогда не бывает одновременно больше одного экземпляра.
//  Создание нового экземпляра с помощью метода start: влечет разрушение старого.
//  Хранит указатель на лабиринт и игровые данные, которые изменяются во время прохождения уровня. 

namespace RoboLabWP.Model.Game
{
    public class MZGame
    {
        private static MZGame instance;

        public MZMaze _maze;
        public MZPosition _currentPosition;
        public MZVisitsMatrix _visitsMatrix;

        int _stepsCount;
        int _minimalSteps;

        bool _isPaused;
        bool _isEnded;

        private MZGame()
        {
            _maze = new MZMaze();
            
            _stepsCount = 0;
            _minimalSteps = _maze.minimalSteps();
            
            _currentPosition = _maze.startingPosition();
            _visitsMatrix = new MZVisitsMatrix(_maze.width(), _maze.height());
            
            visitPosition(_currentPosition);
            _isEnded = false;
            _isPaused = false;
        }

        public static MZGame getInstance()
        {
            if(instance == null)
                {
                    instance = new MZGame();
                }
                return instance;
        }

        public static MZGame Start()
        {
            return instance = new MZGame();
        }

        MZMotion  updateWithFrame(int frame)
        {
            List<MZFlow> flowList = _maze.flowList();
            MZMotion result = null;

            for (int i = 0; i < flowList.Count(); i++)
            {
                result = flowList[i].moveInFlowWithPositionandFrame(_currentPosition,frame);
            }
            return result;
        }

        public MZMoveResultCode makeStepTo(MZDirection direction)
        {
            if (_isPaused)
                return MZMoveResultCode.GAME_PAUSED;

            /*
            List<MZFlow> flowList = _maze.flowList();
            for (int i = 0; i < flowList.Count(); i++)
            {
                if (flowList[i].isInFlow(_currentPosition))
                {
                    if ((direction == flowList[i].direction()) || (direction == (flowList[i].direction() + 2) % 2))
                    {
                        return MZMoveResultCode.MOTION_IMPOSSIBLE;	
                    }
                }
            }*/

            if (!_maze.cells().cellAtPosition(_currentPosition).hasWallAtDirection(direction))
            {
                _currentPosition.move(direction);
                _stepsCount++;
                visitPosition(_currentPosition);
                if ((_maze.winningPosition().x() == _currentPosition.x()) && (_maze.winningPosition().y() == _currentPosition.y()))
                {
                    _isEnded = true;
                    return MZMoveResultCode.GAME_ENDED;
                }
                else
                    return MZMoveResultCode.MOTION_PERFORMED;
            }
               return MZMoveResultCode.MOTION_IMPOSSIBLE;	
        }



        public void visitPosition(MZPosition position)
        {
            _visitsMatrix.setLastVisit(_stepsCount, position);
            MZCellMatrix cellMatrix = _maze.cells();
            MZPosition iteratorPosition = new MZPosition(position.x(), position.y());
            
            while (iteratorPosition.x() > 0)
            {
                if (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.LEFT))
                    break;
                iteratorPosition.move(MZDirection.LEFT);
                _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
            }

            iteratorPosition = new MZPosition(position.x(), position.y());

            while (iteratorPosition.y() > 0)
            {
                if (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.DOWN))
                    break;
                iteratorPosition.move(MZDirection.DOWN);
                _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
            }

            iteratorPosition = new MZPosition(position.x(), position.y());

            while (iteratorPosition.x() < _maze.width() - 1)
            {
                if (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.RIGHT))
                    break;
                iteratorPosition.move(MZDirection.RIGHT);
                _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
            }
            iteratorPosition = new MZPosition(position.x(), position.y());

            while (iteratorPosition.y() < _maze.height() - 1)
            {
                if (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.UP))
                    break;
                iteratorPosition.move(MZDirection.UP);
                _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
            }

            if((!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.LEFT)) && (!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.DOWN)))
            {
                iteratorPosition = new MZPosition(position.x(), position.y());
                iteratorPosition.move(MZDirection.LEFT);
                iteratorPosition.move(MZDirection.DOWN);

                while (iteratorPosition.x() >= 0)
                {
                    if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.RIGHT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.UP))) 
                        break;
                    MZPosition tempPosition = iteratorPosition;
                    while (iteratorPosition.y() >= 0)
                    {
                        if((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.RIGHT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.UP)))
                        {
                            break;
                        }
                        _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
                        iteratorPosition.move(MZDirection.DOWN);
                    }
                    iteratorPosition = tempPosition;
                    iteratorPosition.move(MZDirection.LEFT);
                }
            }

            if ((!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.LEFT)) && (!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.UP)))
            {
                iteratorPosition = new MZPosition(position.x(), position.y());
                iteratorPosition.move(MZDirection.LEFT);
                iteratorPosition.move(MZDirection.UP);

                while (iteratorPosition.x() >= 0)
                {
                    if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.RIGHT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.DOWN)))
                        break;
                    MZPosition tempPosition = iteratorPosition;
                    while (iteratorPosition.y() < _maze.height())
                    {
                        if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.RIGHT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.DOWN)))
                        {
                            break;
                        }
                        _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
                        iteratorPosition.move(MZDirection.UP);
                    }
                    iteratorPosition = tempPosition;
                    iteratorPosition.move(MZDirection.LEFT);
                }
            }

            if ((!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.RIGHT)) && (!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.DOWN)))
            {
                iteratorPosition = new MZPosition(position.x(), position.y());
                iteratorPosition.move(MZDirection.RIGHT);
                iteratorPosition.move(MZDirection.DOWN);

                while (iteratorPosition.x() < _maze.width())
                {
                    if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.LEFT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.UP)))
                        break;
                    MZPosition tempPosition = iteratorPosition;
                    while (iteratorPosition.y() >= 0)
                    {
                        if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.LEFT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.UP)))
                        {
                            break;
                        }
                        _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
                        iteratorPosition.move(MZDirection.DOWN);
                    }
                    iteratorPosition = tempPosition;
                    iteratorPosition.move(MZDirection.RIGHT);
                }
            }

            if ((!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.RIGHT)) && (!cellMatrix.cellAtPosition(position).hasWallAtDirection(MZDirection.UP)))
            {
                iteratorPosition = new MZPosition(position.x(), position.y());
                iteratorPosition.move(MZDirection.RIGHT);
                iteratorPosition.move(MZDirection.UP);

                while (iteratorPosition.x() < _maze.width())
                {
                    if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.LEFT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.DOWN)))
                        break;
                    MZPosition tempPosition = iteratorPosition;
                    while (iteratorPosition.y() < _maze.height())
                    {
                        if ((cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.LEFT)) || (cellMatrix.cellAtPosition(iteratorPosition).hasWallAtDirection(MZDirection.DOWN)))
                        {
                            break;
                        }
                        _visitsMatrix.setLastVisit(_stepsCount, iteratorPosition);
                        iteratorPosition.move(MZDirection.UP);
                    }
                    iteratorPosition = tempPosition;
                    iteratorPosition.move(MZDirection.RIGHT);
                }
            }
        }

        public void pause()
        {
            _isPaused = true;
        }
        public void resume()
        {
            _isPaused = false;
        }
        public bool isPaused()
        {
            return _isPaused;
        }
        public bool isEnded()
        {
            return _isEnded;
        }
        public int stepsCount()
        {
            return _stepsCount;
        }
        public int minimalSteps()
        {
            return _minimalSteps;
        }
    }
}

File: RoboLabWP/RoboLabWPComp/Model/Game/MZGame.cpp
#include "pch.h"
#include "MZGame.h"


MZGame::MZGame(void)
{
	_maze = new MZMaze();
	_currentPosition = _maze->startingPosition();
	_visitsMatrix = new MZVisitsMatrix(_maze->width() +1,_maze->height() + 1);
    _stepsCount = 0;
    visitPosition(_currentPosition);
    _isPaused = false;
    _isEnded = false;
	_minimalSteps = _maze->minimalSteps();
}


MZGame::~MZGame(void)
{
	if (_maze != NULL)
	{
		delete(_maze);
		_maze = NULL;
	}
	if (_visitsMatrix != NULL)
	{
		delete(_visitsMatrix);
		_visitsMatrix = NULL;
	}
	if (_currentPosition != NULL)
	{
		delete(_currentPosition);
		_currentPosition = NULL;
	}
}

MZMotion* MZGame::updateWithFrame( int frame)
{
    vector<MZGameObject*> _objectList = _maze->objectList();
    MZMotion *result = NULL;
    /*for (MZFlow *flow in objectList)
    {
        if(![flow isMemberOfClass: [MZFlow class]])
            continue;
        result = [flow moveInFlowWithPosition:_currentPosition andFrame:frame];
    }*/
	for (unsigned int i = 0; i < _objectList.size(); i++)
	{
		if (_objectList[i] != NULL)
		{
			if (_objectList[i]->info() == FLOW)
			{
				MZFlow *flow = (MZFlow*)_objectList[i];
				result = flow->moveInFlowWithPositionandFrame(_currentPosition, frame);
			}
		}
	}
    return result;
}
int MZGame::stepsCount()
{
    return _stepsCount;
}
int MZGame::minimalSteps()
{
    return _minimalSteps;
}

MZMoveResultCode MZGame::makeStepTo(MZDirection direction)
{
    if(_isPaused)
        return GAME_PAUSED;
    
    vector<MZGameObject*> _objectList = _maze->objectList();
	for (unsigned int i = 0; i < _objectList.size(); i++)
    {
		if (_objectList[i] != NULL)
		{
			if (_objectList[i]->info() == FLOW)
			{
				MZFlow *flow = (MZFlow *)_objectList[i];
				if(flow->isInFlow(_currentPosition))
				{
					if(direction == flow->direction() || direction == ((flow->direction() + 2) % 4))
						return MOTION_IMPOSSIBLE;
				}
			}
        }
		if (_objectList[i] != NULL)
		{
			if (_objectList[i]->info() == ONEWAYWALL)
			{
				MZOneWayWall *oneWayWall = (MZOneWayWall *)_objectList[i];
				if((oneWayWall->position()->x() == _currentPosition->x()) &&
				   (oneWayWall->position()->y() == _currentPosition->y()) &&
				   (oneWayWall->direction() == direction))
					return MOTION_IMPOSSIBLE;
			}
        }
		if (_objectList[i] != NULL)
		{
			if (_objectList[i]->info() == BUTTON)
			{
				MZButton *button = (MZButton *)_objectList[i];
				if((button->position()->x() == _currentPosition->x()) &&
				   (button->position()->y() == _currentPosition->y()) &&
				   (button->direction() == direction))
				{
					button->activate();
					return MZMoveResultCode(MOTION_IMPOSSIBLE | NEEDS_UPDATE); // не уверен
				}
			}
        }
		if (_objectList[i] != NULL)
		{
			if (_objectList[i]->info() == ROTATINGCELL)
			{
				MZRotatingCell *rotatingCell = (MZRotatingCell *)_objectList[i];
				if((rotatingCell->position()->x() == _currentPosition->x()) &&
				   (rotatingCell->position()->y() == _currentPosition->y()) &&
				   (rotatingCell->hasWallAtDirection(direction)))
				{
					return MOTION_IMPOSSIBLE;
				}
			}
        }
    }
    
	if(!_maze->cells()->cellAtPosition(_currentPosition)->hasWallAtDirection(direction))
    {
        _stepsCount++;
        _currentPosition->move(direction);
	    visitPosition(_currentPosition);
        MZPosition *exitPosition;
		for (unsigned int i = 0; i < _objectList.size(); i++)
		{
			if (_objectList[i] != NULL)
			{
				if (_objectList[i]->info() == EXIT)
				{
					MZExit* exit = (MZExit*)_objectList[i];
					exitPosition = exit ->position();
					break;
				}
			}
		}
        if((_currentPosition->x() == exitPosition->x()) && (_currentPosition->y() == exitPosition->y()))
        {
            _isEnded = true;
            return GAME_ENDED;
        }
        else
            return MOTION_PERFORMED;
    }
    return MOTION_IMPOSSIBLE;
}
void MZGame::pause()
{
    _isPaused = true;
}
void MZGame::resume()
{
    _isPaused = false;
}
bool MZGame::isPaused()
{
    return _isPaused;
}
bool MZGame::isEnded()
{
    return _isEnded;
}
void MZGame::visitPosition(MZPosition* position)
{
    _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, position);
    MZCellMatrix *cellMatrix = _maze->cells();
	MZPosition* iteratorPosition = position->copy(); //позиция некоторого абстрактного объекта, которым мы ходим по лабиринту во все стороны по прямой от текущего положения.
    
	while (iteratorPosition->x() > 0)
    {
        if(cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(LEFT))
            break;
        iteratorPosition->move(LEFT);
        _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
    }
	delete(iteratorPosition);
	iteratorPosition = position->copy();
    while (iteratorPosition->y() > 0)
    {
        if(cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(DOWN))
            break;
       iteratorPosition->move(DOWN);
        _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
    }
	delete(iteratorPosition);
    iteratorPosition = position->copy();
	while (iteratorPosition->x() < _maze->width() -1 )
    {
		if(cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(RIGHT))
            break;
		iteratorPosition->move(RIGHT);
        _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
    }
	delete(iteratorPosition);
    iteratorPosition = position->copy();
	while (iteratorPosition->y() < _maze->height() -1 )
    {
		if(cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(UP))
            break;
		iteratorPosition->move(UP);
        _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
    }
    delete(iteratorPosition);
	///

    if((!cellMatrix->cellAtPosition(position)->hasWallAtDirection(LEFT)) &&
       (!cellMatrix->cellAtPosition(position)->hasWallAtDirection(DOWN)))
    {
		iteratorPosition = position->copy();
        iteratorPosition->move(LEFT);
        iteratorPosition->move(DOWN);
        while(iteratorPosition->x() >= 0)
        {
            if( (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(RIGHT)) || 
                (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(UP)))
                break;
            MZPosition *tempPosition = iteratorPosition->copy();
            while (iteratorPosition->y() >= 0)
            {
                if((cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(RIGHT)) ||
                   (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(UP)))
                    break;
                _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
                iteratorPosition->move(DOWN);
            }
			delete(iteratorPosition);
            iteratorPosition = tempPosition->copy();
			delete(tempPosition);
			iteratorPosition->move(LEFT);
			delete(iteratorPosition);
        }
    }

    if((!cellMatrix->cellAtPosition(position)->hasWallAtDirection(LEFT)) &&
       (!cellMatrix->cellAtPosition(position)->hasWallAtDirection(UP)))
    {
		iteratorPosition = position->copy();
        iteratorPosition->move(LEFT);
        iteratorPosition->move(UP);
        while(iteratorPosition->x() >= 0)
        {
            if( (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(RIGHT)) || 
                (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(DOWN)))
                break;
            MZPosition *tempPosition = iteratorPosition->copy();
            while (iteratorPosition->y() >= 0)
            {
                if((cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(RIGHT)) ||
                   (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(DOWN)))
                    break;
                _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
                iteratorPosition->move(UP);
            }
			delete(iteratorPosition);
            iteratorPosition = tempPosition->copy();
			delete(tempPosition);
			iteratorPosition->move(LEFT);
			delete(iteratorPosition);
        }
    }

    if((!cellMatrix->cellAtPosition(position)->hasWallAtDirection(RIGHT)) &&
       (!cellMatrix->cellAtPosition(position)->hasWallAtDirection(DOWN)))
    {
		iteratorPosition = position->copy();
        iteratorPosition->move(RIGHT);
        iteratorPosition->move(DOWN);
        while(iteratorPosition->x() >= 0)
        {
            if( (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(LEFT)) || 
                (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(UP)))
                break;
            MZPosition *tempPosition = iteratorPosition->copy();
            while (iteratorPosition->y() >= 0)
            {
                if((cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(LEFT)) ||
                   (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(UP)))
                    break;
                _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
                iteratorPosition->move(DOWN);
            }
			delete(iteratorPosition);
            iteratorPosition = tempPosition->copy();
			delete(tempPosition);
			iteratorPosition->move(RIGHT);
			delete(iteratorPosition);
        }
    }
 
    if((!cellMatrix->cellAtPosition(position)->hasWallAtDirection(RIGHT)) &&
       (!cellMatrix->cellAtPosition(position)->hasWallAtDirection(UP)))
    {
		iteratorPosition = position->copy();
        iteratorPosition->move(RIGHT);
        iteratorPosition->move(UP);
        while(iteratorPosition->x() >= 0)
        {
            if( (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(LEFT)) || 
                (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(DOWN)))
                break;
            MZPosition *tempPosition = iteratorPosition->copy();
            while (iteratorPosition->y() >= 0)
            {
                if((cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(LEFT)) ||
                   (cellMatrix->cellAtPosition(iteratorPosition)->hasWallAtDirection(DOWN)))
                    break;
                _visitsMatrix->setLastVisitAtStepAtPosition(_stepsCount, iteratorPosition);
                iteratorPosition->move(UP);
            }
			delete(iteratorPosition);
            iteratorPosition = tempPosition->copy();
			delete(tempPosition);
			iteratorPosition->move(RIGHT);
			delete(iteratorPosition);
        }
    }
}